### ISSUE TITLE
[4.0.0.rc2] Memory Leak

### ISSUE THREAD

**Issue by honux** [celery/celery#3205]
Heya.
I have no idea which component exactly is leaking memory, but I found an extremely easy way to reproduce it, that's why I'm reporting here.

The worker
```python
# worker.py
# running with: "celery -A tasks worker --loglevel=info"
from celery import Celery

app = Celery('tasks', broker='redis://127.0.0.1/1', backend='redis://127.0.0.1/2')

@app.task
def add(x, y):
  return x + y
```

The client
```python
# client.py
import time
import random
from celery import Celery, group, signature


if __name__ == "__main__":
  with Celery(broker='redis://127.0.0.1/1', backend='redis://127.0.0.1/2') as celery:
    while True:
      tasks = []
      for i in xrange(random.randint(0, 30)):
        tasks.append(signature('tasks.add', (random.randint(0, 100), random.randint(0, 100))))
      task_group = group(tasks, app=celery)
      result = task_group.apply_async()
      
      time.sleep(0.01)
```

The client will start consuming memory and not releasing it pretty fast. There is no reference to clear / delete / remove anything on the docs.
Tested on Ubuntu 14.04 and 16.04

**Comment by honux** [comment]
New note: Even with the leakage, if you let it run for some hours, it will start consuming HUGE bandwidth (over 50 mbps) and eventually completely stops.

**Comment by ask** [comment]
Thanks! I'm actually aware of this bug, it has something to do with the addition of the `barrier` in celery/canvas.py.  Some circular reference or something is keeping the results alive, but haven't tracked it down yet

**Comment by honux** [comment]
Using the new master, the following code:

```python
from __future__ import absolute_import, unicode_literals
from celery import Celery, chord, group, signature
import time
import random

app = Celery('tests', broker='redis://127.0.0.1/1', backend='redis://127.0.0.1/2')


@app.task(name='Double.Value')
def double_value(value):
  t = random.randint(1, 5)
  return t

@app.task(bind=True, name='Sum.Values')
def sum_values(self, value_1):
  raise self.replace(group(signature('Double.Value', args=(value_1,)), signature('Double.Value', args=(value_1,)), options={'chord': signature('Sum.List')}))

@app.task(name='Sum.List')
def sum_list(value_list):
  return sum(value for value in value_list)
```

with the following client:

```python
from celery import Celery
app = Celery(broker='redis://127.0.0.1/1', backend='redis://127.0.0.1/2')

for t in range(100000):
    t = app.send_task('Sum.Values', args=(10,))
```

makes it leak. way more the old one did :p

**Comment by ask** [comment]
The second leak was added more recently, I fixed this just now in b97e164
